### ISSUE TITLE
Unsupported converter, only named functions, types and lambdas are currently supported with higher-order function

### ISSUE THREAD

**Issue by ericbn** [python/mypy#15736]
**Bug Report**

Mypy succeeds evaluating an attr converter that is an `(Any) -> Any` function, but fails to evaluate a converter that is returned by a higher-order function that returns an `(Any) -> Any` function.

**To Reproduce**

This works as expected:
```python
from attr import define, field


def truncate(val):
    return val[:40]


@define
class Person:
    first_name: str = field(converter=truncate)
    last_name: str = field(converter=truncate)


reveal_type(truncate)
```
Output:
```
scratch.py:14: note: Revealed type is "def (val: Any) -> Any"
Success: no issues found in 1 source file
```

But this does not:
```python
from collections.abc import Callable
from typing import Any

from attr import define, field


def truncate(length) -> Callable[[Any], Any]:
    def truncate_converter(val):
        return val[:length]

    return truncate_converter


@define
class Person:
    first_name: str = field(converter=truncate(40))
    last_name: str = field(converter=truncate(80))


reveal_type(truncate(40))
reveal_type(truncate(80))
```

**Expected Behavior**

Output:
```
scratch.py:20: note: Revealed type is "def (Any) -> Any"
scratch.py:21: note: Revealed type is "def (Any) -> Any"
Success: no issues found in 1 source file
```

**Actual Behavior**

Output:
```
scratch.py:16: error: Unsupported converter, only named functions, types and lambdas are currently supported  [misc]
scratch.py:17: error: Unsupported converter, only named functions, types and lambdas are currently supported  [misc]
scratch.py:20: note: Revealed type is "def (Any) -> Any"
scratch.py:21: note: Revealed type is "def (Any) -> Any"
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.4.1
- Mypy command-line flags: just the file name, no extra flags.
- Mypy configuration options from `mypy.ini` (and other config files):
  ```toml
  [tool.mypy]
  platform = "linux"
  python_version = "3.10"
  ```
- Python version used: 3.10.12

**Comment by ViktorSky** [comment]
I opened one similar to this one in #16860

**Comment by thejcannon** [comment]
We have a spec now: https://github.com/python/typing-council/issues/33

**Comment by filbranden** [comment]
Curious to see whether this feature is coming up.

I've been working on `attrs.Converter` and adding type checker tests for it (e.g. python-attrs/attrs#1373) and it would be great to be able to uncomment the ones that use `field(converter=Converter(...))` and currently trigger this issue.

Thanks!

**Comment by erictraut** [comment]
@filbranden, the typing council signed off on the converter functionality several months ago, and it has been incorporated into the full type checking spec [here](https://typing.readthedocs.io/en/latest/spec/dataclasses.html#converters). The type checker conformance tests have also been updated to cover this functionality (search for `dataclasses_transform_converter` on [this page](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html)).

I don't know if anyone has started working on this support for mypy. I don't see any active PRs. If it helps unblock your type checker test development in `attrs`, pyright has support for this new functionality.

**Comment by filbranden** [comment]
Any chance someone would be able to review #21408, which implements this? I'd be happy to be able to drop `# type: ignore` comments I currently have to work around mypy's current obliviousness to attrs converters. Thanks!
